src/foam/dao/jdbc/MySQLJDBCDAO.java

```diff
@@ -0,0 +1,333 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.dao.jdbc;
+
+import foam.core.*;
+
+import foam.dao.Sink;
+
+import foam.nanos.logger.Logger;
+
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+
+import java.sql.*;
+
+import java.sql.SQLException;
+import java.util.*;
+
+/**
+  We assume that the database is created by a script when the system starts the first time.
+  When the system restarts at any point in time it should verify that the database is already created otherwise it should create it.
+  When creating the database, we only do a CREATE SCHEMA sql instruction, then we create the app user and password and grant him full privileges on this database.
+  Any other database objects (tables) will be created on the fly by the application (when methods of this class are called).
+
+ */
+public class MySQLJDBCDAO extends AbstractJDBCDAO{
+
+  public MySQLJDBCDAO(X x, ClassInfo of, String poolName) throws java.sql.SQLException, ClassNotFoundException {
+    super(x, of, poolName);
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    Connection c = null;
+    IndexedPreparedStatement stmt = null;
+    ResultSet resultSet = null;
+
+    try {
+      c = dataSource_.getConnection();
+      StringBuilder builder = threadLocalBuilder_.get()
+        .append("insert into ")
+        .append(tableName_);
+
+      buildFormattedColumnNames(obj, builder);
+      builder.append(" values");
+      buildFormattedColumnPlaceholders(obj, builder);
+      builder.append(" on duplicate key ")
+        // .append(getPrimaryKey().createStatement()) ... Not in MySQL
+        .append(" update ");
+      buildUpdateFormattedColumnNames(obj, builder);   // Specific to MySQL
+
+      stmt = new IndexedPreparedStatement(c.prepareStatement(builder.toString(),
+        Statement.RETURN_GENERATED_KEYS));
+
+      setStatementValues(stmt, obj);
+
+      int inserted = stmt.executeUpdate();
+      if ( inserted == 0 ) {
+        throw new SQLException("Error performing put_ command");
+      }
+
+      // get auto-generated postgres keys
+/*       resultSet = stmt.getGeneratedKeys();
+      if ( resultSet.next() ) {
+        obj.setProperty(getPrimaryKey().getName(), resultSet.getObject(1));
+      } */
+
+      return obj;
+    } catch (Throwable e) {
+      Logger logger = (Logger) x.get("logger");
+      logger.error(e);
+      return null;
+    } finally {
+      closeAllQuietly(resultSet, stmt);
+    }
+  }
+
+  @Override
+  public FObject remove_(X x, FObject obj) {
```

Comment:
Is this MySQL specific?

Reply:
Apperently no, I will move them to the AbstractJDBCDAO until we find a MySQL specific detail